pkg/reconciler/build/resources/pod.go

```diff
@@ -491,3 +525,470 @@ func getFailureMessage(pod *corev1.Pod) string {
+	ShouldRunPostRun      = true
+	PostRunFile           = "0"
+)
+
+var toolsMount = corev1.VolumeMount{
+	Name:      MountName,
+	MountPath: MountPoint,
+}
+
+var toolsVolume = corev1.Volume{
+	Name: MountName,
+	VolumeSource: corev1.VolumeSource{
+		EmptyDir: &corev1.EmptyDirVolumeSource{},
+	},
+}
+
+type entrypointArgs struct {
+	Args       []string `json:"args"`
+	ProcessLog string   `json:"process_log"`
+	MarkerFile string   `json:"marker_file"`
+
+	ShouldWaitForPrevStep bool   `json:"shouldWaitForPrevStep"`
+	PreRunFile            string `json:"preRunFile"`
+	ShouldRunPostRun      bool   `json:"shouldRunPostRun"`
+	PostRunFile           string `json:"postRunFile"`
+}
+
+// Cache is a simple caching mechanism allowing for caching the results of
+// getting the Entrypoint of a container image from a remote registry. It
+// is synchronized via a mutex so that we can share a single Cache across
+// each worker thread that the reconciler is running. The mutex is necessary
+// due to the possibility of a panic if two workers were to attempt to read and
+// write to the internal map at the same time.
+type Cache struct {
+	mtx   sync.RWMutex
+	cache map[string][]string
+}
+
+// NewCache is a simple helper function that returns a pointer to a Cache that
+// has had the internal cache map initialized.
+func NewCache() *Cache {
+	return &Cache{
+		cache: make(map[string][]string),
+	}
+}
+
+func (c *Cache) get(sha string) ([]string, bool) {
+	c.mtx.RLock()
+	ep, ok := c.cache[sha]
+	c.mtx.RUnlock()
+	return ep, ok
+}
+
+func (c *Cache) set(sha string, ep []string) {
+	c.mtx.Lock()
+	c.cache[sha] = ep
+	c.mtx.Unlock()
+}
+
+type authToken struct {
+	AccessToken string
+	Endpoint    string
+}
+
+type dockerJSON struct {
+	Auths       map[string]authEntry `json:"auths"`
+	HTTPHeaders map[string]string    `json:"HttpHeaders"`
+}
+
+// authEntry is a helper for JSON parsing an "auth" entry of config.json
+// This is not meant for direct consumption.
+type authEntry struct {
+	Auth     string `json:"auth"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Email    string `json:"email,omitempty"`
+}
+
+type dockerCfg map[string]cfgEntry
+
+type cfgEntry struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Email    string `json:"email"`
+	Auth     string `json:"auth"`
+}
+
+type dockerSecret struct {
+	Type                    string `json:"type"`
+	ProjectID               string `json:"project_id"`
+	PrivateKeyID            string `json:"private_key_id"`
+	PrivateKey              string `json:"private_key"`
+	ClientEmail             string `json:"client_email"`
+	ClientID                string `json:"client_id"`
+	AuthURI                 string `json:"auth_uri"`
+	TokenURI                string `json:"token_uri"`
+	AuthProviderX509CertURL string `json:"auth_provider_x509_cert_url"`
+	ClientX509CertURL       string `json:"client_x509_cert_url"`
+}
+
+func getGCRAuthorizationKey() ([]authToken, error) {
```

Comment:
I think `k8schain` should be sufficient for you, but I already implemented this in a way that's consumable by ggcr.

Here's an implementation as a keychain:
https://github.com/google/go-containerregistry/blob/e0f538712eb0767ae6959531febe6041fa466602/pkg/v1/google/keychain.go

Here's an `authn.Authenticator` implementation that uses the same DefaultTokenSource, you can use it with the `remote.WithAuth` option:
https://github.com/google/go-containerregistry/blob/e0f538712eb0767ae6959531febe6041fa466602/pkg/v1/google/auth.go#L45

If you want to get fancy, you could combine your `k8schain` with the `google.Keychain` using `authn.NewMultiKeychain`:
https://github.com/google/go-containerregistry/blob/e0f538712eb0767ae6959531febe6041fa466602/pkg/authn/multikeychain.go

Something like:
```go
// this will first try to authenticate using the k8schain,
// then fall back to the google keychain,
// then fall back to anonymous
mkc := authn.NewMultiKeychain(kc, google.Keychain)
img, _ := remote.Image(ref, remote.WithAuthFromKeychain(kc))
```

Reply:
This is really awesome, this appears to be doing exactly what I wanted for authentication.  Thanks for your help @jonjohnsonjr!!